src/notebooks_service.py

```diff
@@ -132,32 +155,80 @@ def whoami(user):
 
 @app.route(
     urljoin(SERVICE_PREFIX, '<namespace>/<project>/<commit_sha>'),
-    methods=['GET']
+    methods=['GET', 'POST']
 )
 @app.route(
     urljoin(
         SERVICE_PREFIX, '<namespace>/<project>/<commit_sha>/<path:notebook>'
     ),
-    methods=['GET']
+    methods=['GET', 'POST']
 )
 @authenticated
 def launch_notebook(user, namespace, project, commit_sha, notebook=None):
     """Launch user server with a given name."""
     server_name = _server_name(namespace, project, commit_sha)
-
+    notebook_url = _notebook_url(user, server_name, notebook)
     headers = {auth.auth_header_name: 'token {0}'.format(auth.api_token)}
 
-    notebook_url = urljoin(
-        os.environ.get('JUPYTERHUB_BASE_URL'),
-        'user/{user[name]}/{server_name}/'.format(
-            user=user, server_name=server_name
+    status = 'not started'
+
+    if request.method == 'POST':
+        # 0. check if server is already running - if so, redirect us there
+        if server_name in json.loads(
+            requests.request(
+                'GET',
+                '{prefix}/users/{user[name]}'.format(
+                    prefix=auth.api_url, user=user
+                ),
+                headers=headers
+            ).text
+        )['servers']:
+            return redirect(notebook_url)
+
+        # 1. launch using spawner that checks the access
+        payload = {
+            'branch': request.args.get('branch', 'master'),
+            'commit_sha': commit_sha,
+            'namespace': namespace,
+            'notebook': notebook,
+            'project': project,
+        }
+        if os.environ.get('GITLAB_REGISTRY_SECRET'):
+            payload['image_pull_secrets'] = payload.get(
+                'image_pull_secrets', []
+            )
+            payload['image_pull_secrets'].append(
+                os.environ['GITLAB_REGISTRY_SECRET']
+            )
+
+        r = requests.request(
+            'POST',
+            '{prefix}/users/{user[name]}/servers/{server_name}'.format(
+                prefix=auth.api_url, user=user, server_name=server_name
+            ),
+            json=payload,
+            headers=headers,
         )
-    )
 
-    if notebook:
-        notebook_url += '/notebooks/{notebook}'.format(notebook=notebook)
+        # 2. redirect to the server if already running, otherwise send 202
+        if r.status_code == 201:
+            # server is already running
+            app.logger.debug(
+                'server {server_name} running'.format(server_name=server_name)
+            )
+            return redirect(notebook_url)
+
+        elif r.status_code == 202:
+            # server is spawning - return a 202 with a location header
+            app.logger.debug(
+                'spawn initialized for {server_name}'.format(
+                    server_name=server_name
+                )
+            )
+            status = 'spawn initialized'
```

Comment:
is it intended that we don't return here?

Reply:
oops yes this status is superflous. If we return here, it wouldn't make sense to return the html, so I just log that the spawn has been initiated and when we get to the bottom of the function we return with the html that waits for the server.